
# @nolawnchairs/no-import-gaps

This rule enforces having no newlines between import statements.

### ğŸ‘ Invalid
```ts
import foo from 'foo'
import bar from 'bar'

import baz from 'baz'

console.log('Hello, Newman')
```

### ğŸ‘ Valid
```ts   
import foo from 'foo'
import bar from 'bar'
import baz from 'baz'

console.log('Hello, Newman')
```

## Config

The rule takes no arguments.

```ts
{
  '@nolawnchairs/no-import-gaps': 'error'
}
```